### PR TITLE
More OSX glue

### PR DESCRIPTION
**Pull request by kth5**
So I was trying to build the miner on macOS Sierra with both Xcode 8.1 and a homebrewed GCC 6.3 and failed. This PR includes - what I think are - fixes to get it to build properly.

I have tested on a MBP Mid-2015 with an R9 M370X that has 2GB dedicated VRAM but still can't get the miner to start for it fails to build the CL kernel it seems.

With 8 / 1000 I get:
> Error CL_INVALID_BUFFER_SIZE when calling clCreateBuffer to create hash scratchpads buffer.

With 8/ 256 I get:
> Error CL_BUILD_PROGRAM_FAILURE when calling clBuildProgram.

Then again, maybe this PR helps someone to debug those issues. Let me know if there's anything I can try. Can potentially also give temporary access to a Mac with AMD graphics for debugging.

### CODE REVIEW

**Comment by fireice-uk** [comment]
There should be build log error message.

**Comment by kth5** [comment]
This is unfortunately all I get with verbosity on 4:
image

My config:
> "gpu_thread_num" : 1,
> "gpu_threads_conf" : [
>	{ "index" : 0, "intensity" : 10, "worksize" : 8, "affine_to_cpu" : false }
>],
>"platform_index" : 0,
>"use_tls" : false,
>"tls_secure_algo" : true,
>"tls_fingerprint" : "",
>"pool_address" : "some.stratum.pool:1234",
>"wallet_address" : "",
>"pool_password" : "x",
>"call_timeout" : 10,
>"retry_time" : 10,
>"giveup_limit" : 0,
>"verbose_level" : 4,
>"h_print_time" : 60,
>"output_file" : "",
>"httpd_port" : 0,
>"prefer_ipv4" : true,

If there's another quick way to increase verbosity I'm not currently aware but I'm already diving into more detail.

**Comment by fireice-uk** [comment]
Hi, nice to see someone prepared to fix a PR :+1: I pushed an update - there was a 1024 character limit that was in the way of printing a build log. If you rebase you should be able to see what the error is (in all likelyhood something caused by the old gfx card, which will be probably unfixable unfortunately) 

There are other issues that would need changing before I can just merge it (rather than reworking it)

- config.txt needs to stay tracked, otherwise people will need to write their own
- I'm fairly sure CL_INVALID_PIPE_SIZE and CL_INVALID_DEVICE_QUEUE will exist in the right opencl version on MacOS too. Is CL_VERSION_2_0 defined in your headers?

**Comment by kth5** [comment]
Thanks for the log fix, it's now handing out quite a bit more information. :)

>`Build log:
<program source>:18:26: warning: unknown OpenCL extension 'cl_amd_media_ops2' - ignoring
#pragma OPENCL EXTENSION cl_amd_media_ops2 : enable
                         ^
In file included from <program source>:20:
/Users/kth5/code/miners/xmr-stak-amd/./opencl/wolf-aes.cl:76:7: warning: no previous prototype for function 'AES_Round'
uint4 AES_Round(const __local uint *AES0, const __local uint *AES1, const __local uint *AES2, const __local uint *AES3, const uint4 X, const uint4 key)
      ^
/Users/kth5/code/miners/xmr-stak-amd/./opencl/wolf-aes.cl:79:14: warning: implicit declaration of function 'amd_bfe' is invalid in C99
        Y.s0 = AES0[BYTE(X.s0, 0)] ^ AES1[BYTE(X.s1, 1)] ^ AES2[BYTE(X.s2, 2)] ^ AES3[BYTE(X.s3, 3)];
             ^
/Users/kth5/code/miners/xmr-stak-amd/./opencl/wolf-aes.cl:74:21: note: expanded from macro 'BYTE'
#define BYTE(x, y)      (amd_bfe((x), (y) << 3U, 8U))
                    ^
In file included from <program source>:21:
/Users/kth5/code/miners/xmr-stak-amd/./opencl/wolf-skein.cl:31:7: warning: no previous prototype for function 'SKEIN_ROT'
ulong SKEIN_ROT(const uint2 x, const uint y)
      ^
/Users/kth5/code/miners/xmr-stak-amd/./opencl/wolf-skein.cl:33:29: warning: implicit declaration of function 'amd_bitalign' is invalid in C99
        if(y < 32) return(as_ulong(amd_bitalign(x, x.s10, 32 - y)));
                            ^
/System/Library/Frameworks/OpenCL.framework/Versions/A/lib/clang/3.2/include/cl_kernel.h:4274:43: note: expanded from macro 'as_ulong'
#define as_ulong( _x )  __builtin_astype( _x, ulong )
                                          ^
In file included from <program source>:21:
/Users/kth5/code/miners/xmr-stak-amd/./opencl/wolf-skein.cl:33:20: error: invalid reinterpretation: sizes of 'ulong' (aka 'unsigned long') and 'int' must match
        if(y < 32) return(as_ulong(amd_bitalign(x, x.s10, 32 - y)));
                   ^        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/System/Library/Frameworks/OpenCL.framework/Versions/A/lib/clang/3.2/include/cl_kernel.h:4274:25: note: expanded from macro 'as_ulong'
#define as_ulong( _x )  __builtin_astype( _x, ulong )
                        ^
In file included from <program source>:21:
/Users/kth5/code/miners/xmr-stak-amd/./opencl/wolf-skein.cl:34:14: error: invalid reinterpretation: sizes of 'ulong' (aka 'unsigned long') and 'int' must match
        else return(as_ulong(amd_bitalign(x.s10, x, 32 - (y - 32))));
             ^        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/System/Library/Frameworks/OpenCL.framework/Versions/A/lib/clang/3.2/include/cl_kernel.h:4274:25: note: expanded from macro 'as_ulong'
#define as_ulong( _x )  __builtin_astype( _x, ulong )
                        ^
In file included from <program source>:21:
/Users/kth5/code/miners/xmr-stak-amd/./opencl/wolf-skein.cl:37:6: warning: no previous prototype for function 'SkeinMix8'
void SkeinMix8(ulong4 *pv0, ulong4 *pv1, const uint rc0, const uint rc1, const uint rc2, const uint rc3)
     ^
/Users/kth5/code/miners/xmr-stak-amd/./opencl/wolf-skein.cl:47:8: warning: no previous prototype for function 'SkeinEvenRound'
ulong8 SkeinEvenRound(ulong8 p, const ulong8 h, const ulong *t, const uint s)
       ^
/Users/kth5/code/miners/xmr-stak-amd/./opencl/wolf-skein.cl:68:8: warning: no previous prototype for function 'SkeinOddRound'
ulong8 SkeinOddRound(ulong8 p, const ulong8 h, const ulong *t, const uint s)
       ^
/Users/kth5/code/miners/xmr-stak-amd/./opencl/wolf-skein.cl:89:8: warning: no previous prototype for function 'Skein512Block'
ulong8 Skein512Block(ulong8 p, ulong8 h, ulong h8, const ulong *t)
       ^
<program source>:59:6: warning: no previous prototype for function 'keccakf1600'
void keccakf1600(ulong *s)
     ^
<program source>:119:6: warning: no previous prototype for function 'keccakf1600_1'
void keccakf1600_1(ulong *st)
     ^
<program source>:178:6: warning: no previous prototype for function 'keccakf1600_2'
void keccakf1600_2(ulong *st)
     ^
<program source>:306:6: warning: no previous prototype for function 'CNKeccak'
void CNKeccak(ulong *output, ulong *input)
     ^
<program source>:330:26: warning: unknown OpenCL extension 'cl_amd_media_ops2' - ignoring
#pragma OPENCL EXTENSION cl_amd_media_ops2 : enable
                         ^
<program source>:336:6: warning: no previous prototype for function 'AESExpandKey256'
void AESExpandKey256(uint *keybuf)
     ^`

**Comment by kth5** [comment]
So I dug around a bit and couldn't find amd_bitalign or the cl_amd_media_ops2 extension on OSX, which is used in wolf-skein.cl:33 and only part of a more recent version of AMD's APP SDK (3.0 on Linux here has it) which coincidentally support OpenCL 2.0 as well.

Thus, unless someone feels inclined to port the kernels to OSX OpenCL 1.2 - which I think is unreasonable to ask for - then this PR is only of questionable use. Then again, I could be wrong and the fix is easy. Just haven't wrapped my head around it yet.

**Comment by fireice-uk** [comment]
I will be very busy tomorrow, but I should have some time to look into it during the weekend.

**Comment by kth5** [comment]
That would be amazing indeed! Just let me know if you can use me to get the necessary information. Until then I'll study up on the OpenCL parts and see if I can tweak a fix in myself. :)

**Comment by fireice-uk** [comment]
Just an update, it is possible to tweak it to not use the semi-new amd extensions, however I'm fairly short on time right now. If you want to play around with it - you are more than welcome to.

You will need to implement two functions used from the extensions amd_bitalign and amd_bfe

**Comment by kth5** [comment]
Already been looking into it. I'm not very proficient with OpenCL so we'll see how far I'll get. Also working with time constraints here. ;)

**Comment by kth5** [comment]
Using the reference docs I tried to wrap my head around amd_bfe after finding a neat example for amd_bitalign. I haven't been able to verify it works for there are a few remaining errors curing OpenCL kernel compilation. Attaching log:

[log.txt](https://github.com/fireice-uk/xmr-stak-amd/files/877773/log.txt)

**Comment by kth5** [comment]
Last two commits got the kernel to compile, guess I had the return type wrong. Anyway, now the CL kernel builds but during execution throws the following:

image

Will investigate this further in a while, it may still be a problem caused by the two emulated functions introduced in this PR.

**Comment by fireice-uk** [comment]
It is more likely to be a problem with the worksize - have you tried setting it to 1? If that doesn't work try setting the second argument on reqd_work_group_size attributes to 1 too.

**Comment by kth5** [comment]
So with a worksize of 1 and an intensity of 40 I seem to be hitting the sweet spot on this R9 M370X (2GB). 25H/s... well, not worth it but it seems to be hashing away. I'll leave the miner running for a few hours to verify correctness.

**Comment by kth5** [comment]
So, getting a GPU compute error whenever a share is found (?). Other than that it seems to be happily hashing and detecting new blocks. Not very useful yet indeed.

I'm betting it's the implementation of amd_bfe of mine which is amateurish at best. I'll have another shot at it when I find more time.

**Comment by fireice-uk** [comment]
That would cause it. Shares are double checked on CPU, if your algo is wrong then all of them will be incorrect.

**Comment by fireice-uk** [comment]
Another thing - nicehash has special nonce requirements. Double check on a kosher pool.

**Comment by fireice-uk** [comment]
If you want to play around some more I don't mind guiding you. This is the line where the performance is breaking:
https://github.com/fireice-uk/xmr-stak-amd/blob/master/opencl/wolf-aes.cl#L74

Try doing it in other way such as an and mask and shift.

File: amd_gpu/gpu.c [hunk 1]
```diff
@@ -163,10 +163,12 @@ const char* err_to_str(cl_int ret)
 		return "CL_INVALID_LINKER_OPTIONS";
 	case CL_INVALID_DEVICE_PARTITION_COUNT:
 		return "CL_INVALID_DEVICE_PARTITION_COUNT";
+#ifndef __APPLE__
```
Comment (psychocrypt) [inline amd_gpu/gpu.c]: This guard is not only a fix for APPLE it fixes also `ubuntu 14.04`. My suggestion is to check the AMD APP SDK VERSION and disable this code for old APP SDK's

Comment (kth5) [inline amd_gpu/gpu.c]: We'd need to do both as there is no direct support of AMD for Apple's OS. We need to deal with Apple's SDK which seems to be an entirely different thing.

I'll investigate checking for AMD APPSDK version for you on one of my Ubuntu boxes.

Comment (psychocrypt) [inline amd_gpu/gpu.c]: I fixed it in #15 because this is an issue with OpenCl < 2.0 and independent of the APPSDK.

Comment (kth5) [inline amd_gpu/gpu.c]: Unfortunately I won't be able to spend too much time on this in the near future so the PR may end up pulling master a few times anyway.

**Comment by psychocrypt** [comment]
Could you try to build this branch on top of #13, #15 and #16 

I fixed a bug in #13 and refactor cmake in #16 thus all should be more platform independent.

**Comment by psychocrypt** [comment]
I read that you have not time for a rebase. I will put all your changes on top of my after they are merged, after that you could test if it is compiling on MAC.

**Comment by kth5** [comment]
Merged your changes and it still builds like a charm. I removed Apple specifics where applicable too. Performs as before.

**Comment by psychocrypt** [comment]
@fireice-uk Opened a dev branch to simplify releases and and support to test non stable changes, I will create a new pull request (in your name) to the dev on top of all my changes, because this pull  also contains some config changes. 

thx for testing!!

File: config.txt [hunk 2]
```diff
@@ -11,12 +11,7 @@
  * affine_to_cpu - This will affine the thread to a CPU. This can make a GPU miner play along nicer with a CPU miner.
  */
 "gpu_threads_conf" : [ 
-	{ "index" : 0, "intensity" : 1000, "worksize" : 8, "affine_to_cpu" : false },
-	{ "index" : 1, "intensity" : 1000, "worksize" : 8, "affine_to_cpu" : false },
-	{ "index" : 2, "intensity" : 1000, "worksize" : 8, "affine_to_cpu" : false },
-	{ "index" : 3, "intensity" : 1000, "worksize" : 8, "affine_to_cpu" : false },
-	{ "index" : 4, "intensity" : 1000, "worksize" : 8, "affine_to_cpu" : false },
-	{ "index" : 5, "intensity" : 1000, "worksize" : 8, "affine_to_cpu" : false },
+	{ "index" : 0, "intensity" : 16, "worksize" : 1, "affine_to_cpu" : false }
```
Comment (psychocrypt) [inline config.txt]: btw Wit my changes it should now possible to change the worksize `1` to `8`

File: opencl/cryptonight.cl [hunk 3]
```diff
@@ -13,7 +13,36 @@
...
+#else
+#pragma "Emulating amd_bitalign"
+// taken from https://github.com/Bdot42/mfakto/blob/master/src/common.cl
+// we should define something for bitalign() ...
+//     Build-in Function
+//      uintn  amd_bitalign (uintn src0, uintn src1, uintn src2)
+//    Description
+//      dst.s0 =  (uint) (((((long)src0.s0) << 32) | (long)src1.s0) >> (src2.s0 & 31))
+//      similar operation applied to other components of the vectors.
+#define amd_bitalign(src0, src1, src2) (src0  << (32-src2)) | (src1 >> src2)
```
Comment (psychocrypt) [inline opencl/cryptonight.cl]: BUG: the parentheses are wrong placed and missing cast to 64 bit: this is solve din #17

**Comment by psychocrypt** [comment]
I will close this PR all needed changes are now in #17 and the BUGs are fixed.

**Pull request closed without merging**